src/serac/physics/utilities/finite_element_state.cpp

```diff
@@ -9,16 +9,97 @@
 namespace serac {
 
 FiniteElementState::FiniteElementState(mfem::ParMesh& mesh, FiniteElementState::Options&& options)
-    : mesh_(mesh),
+    : mesh_(&mesh),
       coll_(options.coll ? std::move(options.coll)
                          : std::make_unique<mfem::H1_FECollection>(options.order, mesh.Dimension())),
-      space_(&mesh, coll_.get(), options.vector_dim, options.ordering),
-      gf_(std::make_unique<mfem::ParGridFunction>(&space_)),
-      true_vec_(&space_),
+      space_(
+          std::make_unique<mfem::ParFiniteElementSpace>(&mesh, &retrieve(coll_), options.vector_dim, options.ordering)),
+      // Leave the gridfunction unallocated so the allocation can happen inside the datastore
+      // Use a raw pointer here, lifetime will be managed by the DataCollection
+      gf_(new mfem::ParGridFunction(&retrieve(space_), static_cast<double*>(nullptr))),
+      true_vec_(&retrieve(space_)),
       name_(options.name)
 {
-  *gf_      = 0.0;
   true_vec_ = 0.0;
 }
 
+FiniteElementState::FiniteElementState(mfem::ParMesh& mesh, mfem::ParGridFunction& gf, const std::string& name)
```

Comment:
A large comment describing why the `MaybeOwned` type and the general memory management strategy of restart vs. non-restart runs would be useful.